---
title: "12SV5 reference"
output:
  pdf_document:    toc: true
  html_document:
    df_print: paged
    toc: true
    toc_float: true
---

Stepping through this notebook will build a *12SV5* reference database, trimmed to sequences amplified by the desired primer set.

This notebook use
```{r libraries, include=FALSE, echo=FALSE}
library(Biostrings)
library(here)
library(ShortRead) # for clean()
library(taxonomizr) # for NCBI accession lookup
library(tidyverse)

source(here('code', 'functions', 'find_primer_pair.R'))
source(here('code', 'functions', 'query_ncbi.R'))
```

```{r plotting, include=FALSE, echo=FALSE}
theme_set(theme_bw() +
               theme(
                    axis.text = element_text(size = 12),
                    axis.title = element_text(size = 14,
                                              face = 'bold'),
                    legend.title = element_text(size = 12,
                                                face = 'bold'),
                    strip.text = element_text(size = 12,
                                              face = 'bold')
                     )
)
```

# Read in data

```{r}
# Primer sequences
V5F <- DNAString('TAGAACAGGCTCCTCTAG')
V5R <- DNAString('TTAGATACCCCACTATGC')

primers <- list(V5F, V5R)
```

```{r}
# Manually curated list of dietary and medicinal plants
animals <- 
     here('data', 'inputs', 'human-foods.csv') |>
     read.csv(stringsAsFactors = FALSE) |>
     filter(category == 'animal') |> 
     pull(scientific_name)

length(animals)
head(animals)
tail(animals)
```

```{r}
# SQL reference
sql <- here('accessionTaxa.sql')
```

```{r}
# Parsed RefSeq data (last organized Jan 2023)
mito <- 
     readDNAStringSet(
          here('data', 
               'outputs', 
               'parsed-refs',
               'RefSeq',
               'refseq_mito_species.fasta'))

mito
```

# Submit query

## RefSeq (local)

### Find primers

```{r}
# Note that there are lots of sequences that include Ns
length(mito)
length(clean(mito))
```

```{r}
refseq.12SV5 <- find_primer_pair(mito, 
                                fwd = primers[[1]],
                                rev = primers[[2]])

cat(length(refseq.12SV5), 'sequences have the primer set')
```

### Subset to foods

Just doing this by a simple grep for now
Could imagine it would be cleaner to process RefSeq to a comparable name as in our query
Keep accessions from raw files, look up taxonomy, then assign a "lowest level" name that would correspond to entries in human-foods.txt.

```{r}
# Find indices of entries matching 
animals.i <- 
     lapply(animals, grep, x = names(refseq.12SV5)) %>%
     unlist()

cat('There are', length(animals), 'food animals in our query\n')

# Subset
refseq.12SV5 <- refseq.12SV5[animals.i]
cat(length(refseq.12SV5), 'have a 12SV5 sequence in the RefSeq mito database')
```

```{r}
# Strip name to only NCBI accession
names(refseq.12SV5) %>% head()
								  

names(refseq.12SV5) <- 
     gsub(names(refseq.12SV5),
          pattern = ' .*$',
          replacement = '')

head(names(refseq.12SV5))
	
```

## NCBI (remote)

```{r}
# Pull sequences from NCBI
ncbi.12SV5 <- query_ncbi(marker = '12S',
                        organisms = animals)
```

This is from the total number of available sequences

```{r}
length(ncbi.12SV5)
length(clean(ncbi.12SV5))
```

### Find primers
Now look for primer binding sites within retrieved sequences.

```{r}
ncbi.12SV5 <- find_primer_pair(ncbi.12SV5, 
                              fwd = primers[[1]],
                              rev = primers[[2]])

cat(length(ncbi.12SV5), 'sequences have the primer set')
```

```{r}
# Note some entries are marked as unverified
names(ncbi.12SV5)[grepl('UNVERIFIED', names(ncbi.12SV5))] |> 
     head(5)

# Remove 
length(ncbi.12SV5)
ncbi.12SV5 <- ncbi.12SV5[!(grepl('UNVERIFIED', names(ncbi.12SV5)))]
length(ncbi.12SV5)
```
Want to convert these long, descriptive names to just an accession and taxon. Strip to accession number, and then use this to do a taxonomic lookup.

```{r}
# Strip name to only NCBI accession
names(ncbi.12SV5) |> head()
names(ncbi.12SV5) <- 
     names(ncbi.12SV5) |> 
     gsub(pattern = ' .+$', replacement = '') |> 
     gsub(pattern = '^>', replacement = '')

head(names(ncbi.12SV5))
```

## Combine

### Check overlap

```{r}
length(refseq.12SV5)
length(ncbi.12SV5)
```

```{r}
# Named as accession numbers:
intersect(names(ncbi.12SV5), names(refseq.12SV5)) |> length()
setdiff(names(refseq.12SV5), names(ncbi.12SV5)) |> length()
setdiff(names(ncbi.12SV5), names(refseq.12SV5)) |> length()
```

Theoretically, RefSeq is entirely contained within NCBI's nucleotide record, but there are entries that are unique to RefSeq here.  Think this is because I restrict the query to "big" NCBI to have the term "12S" in the record name-- not impossible to overcome, but currently don't have a strategy for handling the # and length of records that get pulled down without that filter in our query term.  This can be an area for future updates.

### Merge
```{r}
# Data frame of results
seqs.df <- 
     data.frame(source = 'RefSeq',
                accession = names(refseq.12SV5),
                seq = as.character(refseq.12SV5))

seqs.df <- 
     data.frame(source = 'GenBank',
                accession = names(ncbi.12SV5),
                seq = as.character(ncbi.12SV5)) |> 
     bind_rows(seqs.df)

head(seqs.df)
```

### Taxonomy

```{r}
# Look up accession taxonomy using taxonomizr-formatted SQL database
ids <- taxonomizr::accessionToTaxa(seqs.df$accession, sql)
```

```{r}
# Any names missing?
any(is.na(ids))
sum(is.na(ids))
```

```{r}
# Which ones?
missing.df <- seqs.df[is.na(ids), c('source', 'accession')]
missing.df
```

Missing entries are sequence records that have been added to NCBI in the time between making the taxonomic SQL file and now. Note the output of the previous chunk. The first column describes the index of ids, the third column is the accession ID, which you will manually query in the NCBI website. From the result, find the taxon ID. This will be the value at the right-hand side below.
```{r}
source(here('code', 'functions', 'query_ncbi_accession.R'))

for(i in seq_len(nrow(missing.df))) {
     idx <- rownames(missing.df)[i]
     idx <- as.integer(idx)
     acc <- missing.df[i,2]
     ids[idx] <- query_ncbi_accession(acc)
}

# Manually add these by querying in browser with accession
# IMPORTANT: If repeating, check these to make sure they still line up
# If any additions/changes to input files, likely they will not
#ids[5742] <- 3483 # NC_027223.1 Cannabis sativa  

# Remove entries not in query_ncbi_accession
which(is.na(ids))
ids <- na.omit(ids)

# Confirm got them all
any(is.na(ids))
```
```{r}
taxonomy.raw <- taxonomizr::getRawTaxonomy(ids, sql)
```

```{r}
# Pull desired levels from this structure
# Not working within getTaxonomy function
vars <- c("superkingdom", 
          "phylum", 
          "class", 
          "order", 
          "family", 
          "genus",
          "species",
          "subspecies")

taxonomy <- data.frame(superkingdom = NULL,
                       phylum = NULL,
                       class = NULL,
                       order = NULL,
                       family = NULL,
                       genus = NULL,
                       species = NULL,
                       subspecies = NULL)

for (i in seq_along(taxonomy.raw)){
     row.i <- 
          taxonomy.raw[[i]] |> 
          t() |> 
          data.frame() 
     
     # Pick columns we're interested in
     shared <- intersect(vars, names(row.i))
     row.i <- select(row.i, one_of(shared))
     
     taxonomy <- bind_rows(taxonomy, row.i)
}

# Add taxon ID
taxonomy$taxid <- 
     names(taxonomy.raw) |> 
     trimws() |> 
     as.integer()

taxonomy <- select(taxonomy, taxid, everything())
```

```{r}
head(taxonomy)
```

```{r}
# Join back to accession
nrow(taxonomy) == nrow(seqs.df)

seqs.df <- 
     bind_cols(seqs.df,
               taxonomy)
```

```{r}
# Get lowest-level taxon name
seqs.df <- 
     seqs.df |> 
     MButils::lowest_level() |>
     rename(taxon = 'name') |> 
     select(source, accession, taxon, taxid, superkingdom:subspecies, seq)
```


#### QC

Check for common errors

```{r}
# Check for degenerate nucleotide characters
grep('[AGCT]*[^AGCT]+', seqs.df$seq)
```

```{r}
# Add a flag to these taxa, to see if there's a back-up sequence
seqs.df$N <- grepl('[AGCT]*[^AGCT]+', seqs.df$seq)

with_n <- 
     seqs.df$taxon[grepl('[AGCT]*[^AGCT]+', seqs.df$seq)] |> 
     unique()
```

```{r}
seqs.df |> 
     filter(taxon %in% with_n) |> 
     group_by(taxon, N) |> 
     count() |> 
     ungroup() |> 
     group_by(taxon) |> 
     summarize(any(!N)) |> 
     arrange(`any(!N)`)
```
Okay, so for all taxa but
- Amaranthus cruentus
- Amaranthus dubius
- Costus erythrophyllus
- Vaccinium tenellum
- Victoria amazonica

we are covered by a second sequence. 

```{r}
# Remove sequences containing Ns
seqs.df <- 
     filter(seqs.df,
            !grepl(pattern = '[AGCT]*[^AGCT]+', seq))
```

####  Get reads in same orientation

```{r}
# Get orientation of sequence by finding primers
# How many mismatches are allowed?
fwd_err <- floor(0.2*length(V5F))
rev_err <- floor(0.2*length(V5R))

fwd_err
rev_err
```

```{r}
ref <- DNAStringSet(seqs.df$seq)
names(ref) <- paste(seqs.df$accession, seqs.df$taxon)
```

```{r}
# Forward primer at start of read
fwd_matches <- 
     vmatchPattern(V5F, 
                   ref, 
                   max.mismatch = fwd_err,
                   fixed = TRUE) |>
     as.data.frame() |> 
     filter(start <= 1) |> 
     mutate(type = 'forward') |> 
     select(group, type)

# Reverse primer at start of read
rev_matches <- 
     vmatchPattern(V5R, 
                   ref, 
                   max.mismatch = rev_err,
                   fixed = TRUE) |>
     as.data.frame() |> 
     filter(start <= 1) |> 
     mutate(type = 'reverse') |> 
     select(group, type)
```

```{r}
seqs.df <- 
     bind_rows(fwd_matches,
          rev_matches) |> 
     arrange(group) |> 
     bind_cols(seqs.df)

seqs.df |> 
     group_by(type) |> 
     count()
```

```{r}
nrow(rev_matches) + nrow(fwd_matches) == nrow(seqs.df)
```

Now, if a read is reversed, we want to replace it with its reverse complement
```{r}
seqs.df <- 
     mutate(seqs.df,
            seq = ifelse(type == 'reverse',
                         yes = seq |> 
                              DNAStringSet() |> 
                              reverseComplement() |> 
                              as.character(),
                         no = seq)) |> 
     select(-c(type, group))
```

#### Clean results

* Sequences that are the same and that come from the same species can be de-duplicated
* Sequences that are different and come from the same species must be preserved

```{r}
seqs.df |> 
     group_by(taxon, seq)
```
```{r}
dups <- 
     seqs.df |> 
     group_by(taxon) |> 
     summarize(n = sum(duplicated(seq)))

arrange(dups, desc(n))
```

```{r}
sum(dups$n)
```

So the number of sequences we expect after filtering is 
```{r}
dim(seqs.df)[1] - sum(dups$n)
```

```{r}
seqs.df <- 
     seqs.df |>
     group_by(superkingdom,
              phylum,
              class,
              order,
              family,
              genus,
              species,
              subspecies,
              taxon, 
              seq) |> 
     arrange(desc(source), accession) |> # Puts RefSeq accessions first 
     summarize(accession = first(accession)) # Choose the first accession number

dim(seqs.df)
```

## Save

### DADA2

```{r}
# Sort alphabetically (first by species name, and then accession number)
seqs.df <- arrange(seqs.df, 
                   taxon,
                   accession)

# Remove entries that do not have taxonomy
seqs.df <- seqs.df[!is.na(seqs.df$superkingdom),]

# Convert back to DNAStringSet object
s12SV5 <- seqs.df$seq
names(s12SV5) <- paste(seqs.df$accession, seqs.df$taxon)

s12SV5 <- DNAStringSet(s12SV5)
s12SV5
```

```{r}
# Save to file
writeXStringSet(s12SV5,
            here('data',
                 'outputs',
                 'dada2-compatible',
                 '12SV5',
                 '12SV5FR.fasta'))
```

```{r}
seqs.df <- 
     seqs.df |> 
     unite(col = 'name',
           superkingdom:subspecies,
           sep = ';') 

names(s12SV5) <- seqs.df$name
```

```{r}
writeXStringSet(s12SV5,
            here('data',
                 'outputs',
                 'dada2-compatible',
                 '12SV5',
                 '12SV5FR_taxonomy.fasta'))
```

### QIIME2 

Taxonomic assignment in QIIME2 with `classify-consensus-vsearch` requires two files:
- `i-reference-reads`, a QIIME artifact containing a FASTA file of sequences identified by their accession
- `i-reference-taxonomy`, a QIIME artifact containing a TSV (tab-separated value) file with two columns: "Feature ID", the same accessions, and "Taxon", the taxonomic lineage with ranks separated by semicolons

```{r}
outdir <- here('data',
               'outputs',
               'qiime2-compatible')

Sys.setenv(QIIME2 = outdir)
```

Note that QIIME2 requires that the FASTA file have unique accessions, so make these unique before proceeding

```{r}
seqs.df <- mutate(seqs.df,
                  accession = make.unique(accession,
                                    sep = '_'))

names(s12SV5) <- seqs.df$accession
writeXStringSet(s12SV5,
            file.path(outdir,
            '12SV5-sequences.fasta'))
```

```{r}
seqs.df |>
 select(`Feature ID` = accession,
        Taxon = name) |>
 write_delim(delim = '\t',
             file.path(outdir,
                       '12SV5-taxonomy.tsv'))
```

```{r}
sessionInfo()
```